### PR TITLE
EICNET-1858: group ownership transfer confirmation (rework expiration)

### PR DESCRIPTION
**Pull request by 4hardfire**
### Improvements

- Create new message types to send notification when a request to transfer ownership has been expired;
- Create cron task to close expired requests and send notifications;
- Do not allow users to accept/deny expired request;
- Show information about expired request when viewing list of group members.

### Tests

- [ ] When requesting a new ownership transfer, make sure there is a new field to provide the timeout for the request to be automatically closed
- [ ] When request has expired, make sure you see a message in the list of members bellow the user that was requested
- [ ] Try to access the accept/deny links and make sure you get an access denied
- [ ] Run cron and make sure the request was automatically closed and you get 2 notifications (1 for the author of the request and another one for the user who didn't accept the request)

### CODE REVIEW

File: lib/modules/eic_flags/src/Hooks/CronOperations.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          $query->join('flagging__field_request_timeout', 'frt', 'f.id = frt.entity_id');
+          $query->join('flagging__field_request_status', 'frs', 'f.id = frs.entity_id');
+          $query->condition('f.flag_id', $flag_id);
+          $query->condition('frs.field_request_status_value', RequestStatus::OPEN);
+          $query->condition('frt.field_request_timeout_value', 0, '>');
+          $query->where($now->getTimestamp() . ' >= ((frt.field_request_timeout_value * 86400) + f.created)');
+          $query->fields('f', ['id', 'flag_id', 'entity_type', 'entity_id']);
+          $result = $query->execute()->fetchAll();
+          foreach ($result as $row) {
+            $flag = $this->entityTypeManager->getStorage('flagging')
```
Comment (Dromadoros) [inline lib/modules/eic_flags/src/Hooks/CronOperations.php:121]: Maybe just add doc for both variables (flag, flagged_entity) so phpstorm does not throw warning on the method request TimeOut

```
            /** @var \Drupal\flag\FlaggingInterface $flag */
            $flag = $this->entityTypeManager->getStorage('flagging')
              ->load($row->id);
            /** @var \Drupal\Core\Entity\ContentEntityInterface $flagged_entity */
            $flagged_entity = $this->entityTypeManager->getStorage($row->entity_type)
              ->load($row->entity_id);
```

File: lib/modules/eic_flags/src/Hooks/CronOperations.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager'),
+      $container->get('eic_flags.handler_collector'),
+      $container->get('database'),
+      $container->get('module_handler')
```
Comment (Dromadoros) [inline lib/modules/eic_flags/src/Hooks/CronOperations.php:86]: We are never assigning it, we can remove this line

File: lib/modules/eic_flags/src/Service/AbstractRequestHandler.php [hunk 3]
```diff
@@ -428,4 +440,71 @@ public function getSupportedResponsesForClosedRequests() {
     ];
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function hasExpiration(FlaggingInterface $flag) {
+    $fields = $this->entityFieldManager->getFieldDefinitions('flagging', $flag->getFlagId());
+    if (
```
Comment (Dromadoros) [inline lib/modules/eic_flags/src/Service/AbstractRequestHandler.php:448]: Shorter to use this I guess:

```
    $fields = $this->entityFieldManager->getFieldDefinitions('flagging', $flag->getFlagId());
    
    return isset($fields[HandlerInterface::REQUEST_TIMEOUT_FIELD]) &&
      $flag->get(HandlerInterface::REQUEST_TIMEOUT_FIELD)->value > 0;
```

File: lib/modules/eic_flags/src/Service/AbstractRequestHandler.php [hunk 4]
```diff
@@ -428,4 +440,71 @@ public function getSupportedResponsesForClosedRequests() {
...
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function hasExpired(FlaggingInterface $flag) {
+    $fields = $this->entityFieldManager->getFieldDefinitions('flagging', $flag->getFlagId());
```
Comment (Dromadoros) [inline lib/modules/eic_flags/src/Service/AbstractRequestHandler.php:461]: Better to re-use the hasExpiration method

```
    if (!$this->hasExpiration($flag)) {
      return FALSE;
    }

    $now = DrupalDateTime::createFromTimestamp(time());
    $limit = ($flag->get(HandlerInterface::REQUEST_TIMEOUT_FIELD)->value * 86400) + $flag->get('created')->value;

    return $now->getTimestamp() >= $limit;
```

File: lib/modules/eic_flags/src/Hooks/CronOperations.php (hunk 1 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Hooks/CronOperations.php:121]: Done!

File: lib/modules/eic_flags/src/Hooks/CronOperations.php (hunk 2 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Hooks/CronOperations.php]: Done!

File: lib/modules/eic_flags/src/Service/AbstractRequestHandler.php (hunk 3 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Service/AbstractRequestHandler.php]: Done!

File: lib/modules/eic_flags/src/Service/AbstractRequestHandler.php (hunk 4 above)
Comment (4hardfire) [reply lib/modules/eic_flags/src/Service/AbstractRequestHandler.php]: Done!

**Review by Dromadoros — APPROVED** [review]

**Pull request merged**